sklearn/_loss/loss.py

```diff
@@ -968,6 +1080,94 @@ def predict_proba(self, raw_prediction):
         proba[:, 0] = 1 - proba[:, 1]
         return proba
 
+    def loss_array_api(self, y_true, raw_prediction, sample_weight=None, xp=None):
+        xp, _ = get_namespace(y_true, raw_prediction, sample_weight, xp=xp)
+        return self._compute_loss_array_api(
+            y_true=y_true,
+            raw_prediction=raw_prediction,
+            raw_prediction_exp=xp.exp(raw_prediction),
+            neg_raw_prediction_exp=xp.exp(-raw_prediction),
+            sample_weight=sample_weight,
+            xp=xp,
+        )
+
+    def gradient_array_api(self, y_true, raw_prediction, sample_weight=None, xp=None):
+        xp, _ = get_namespace(y_true, raw_prediction, sample_weight, xp=xp)
+        return self._compute_gradient_array_api(
+            y_true=y_true,
+            raw_prediction=raw_prediction,
+            raw_prediction_exp=xp.exp(raw_prediction),
+            neg_raw_prediction_exp=xp.exp(-raw_prediction),
+            sample_weight=sample_weight,
+            xp=xp,
+        )
+
+    def loss_gradient_array_api(
+        self, y_true, raw_prediction, sample_weight=None, xp=None
+    ):
+        xp, _ = get_namespace(y_true, raw_prediction, sample_weight, xp=xp)
+        raw_prediction_exp = xp.exp(raw_prediction)
+        neg_raw_prediction_exp = xp.exp(-raw_prediction)
+        loss = self._compute_loss_array_api(
+            y_true=y_true,
+            raw_prediction=raw_prediction,
+            raw_prediction_exp=raw_prediction_exp,
+            neg_raw_prediction_exp=neg_raw_prediction_exp,
+            sample_weight=sample_weight,
+            xp=xp,
+        )
+        gradient = self._compute_gradient_array_api(
+            y_true=y_true,
+            raw_prediction=raw_prediction,
+            raw_prediction_exp=raw_prediction_exp,
+            neg_raw_prediction_exp=neg_raw_prediction_exp,
+            sample_weight=sample_weight,
+            xp=xp,
+        )
+        return loss, gradient
+
+    def _compute_loss_array_api(
+        self,
+        y_true,
+        raw_prediction,
+        raw_prediction_exp,
+        neg_raw_prediction_exp,
+        sample_weight=None,
+        xp=None,
+    ):
+        log1pexp = xp.where(
+            raw_prediction <= 18,
+            xp.log1p(raw_prediction_exp),
+            xp.where(
+                raw_prediction <= 33.3,
+                raw_prediction + neg_raw_prediction_exp,
+                raw_prediction,
+            ),
+        )
```

Comment:
The matching Cython code has more levels:

```cython
cdef inline double log1pexp(double x) noexcept nogil:
    if x <= -37:
        return exp(x)
    elif x <= -2:
        return log1p(exp(x))
    elif x <= 18:
        return log(1. + exp(x))
    elif x <= 33.3:
        return x + exp(-x)
    else:
        return x
```

I think reflecting this in the array api version could further align the numerical stability of both branches and might make it possible to make the logistic regression compliance test pass without scaling the data for all float32 cases.

Reply:
I did try implementing more levels but the `loss` tests pass with the current level of nesting so I did not go for more.